lib/omnibus/fetchers/git_fetcher.rb

```diff
@@ -151,9 +150,9 @@ def git_checkout(ref=resolved_version)
     #
     # @return [void]
     #
-    def git_fetch(ref=resolved_version)
-      git("fetch --all")
-      git("reset --hard #{ref}")
+    def git_fetch
+      git("fetch #{source_url} #{described_version}#{' --recurse-submodules=on-demand' if clone_submodules?}")
```

Comment:
Let's try not to be too tricky here. This violates the 80-character line-length, but it's also difficult to understand what is going on. Even though it's longer, I think the following is much clearer and less likely to cause confusion in the future:

```ruby
fetch_cmd = "fetch #{source_url} #{described_version}"
fetch_cmd << " --recurse-submodules=on-demand" if clone_submodules?
git(fetch_cmd)
```

What do you think?